### PR TITLE
Windows: Fix ssl_socket_test

### PR DESCRIPTION
**Pull request by sunjayBhatia**
Commit Message:
Windows: Fix ssl_socket_test

- SmallReadsIntoSameSlice test: use write method on connection instead
  of transport socket so SSL_write failures will be properly retried,
  also otherwise empty writes from connection will trigger failed
  assertions in buffer::linearize
- When SSL shutdown is not appropriate, perform basic socket shutdown;
  relevant when client certificate is invalid and connection should be
  dropped by server, however we still need to ensure TLS alerts are
  able to be read by client, shutdown on socket ensures this can happen
- Replace `\r` with `""` when comparing test fixtures to results to ensure line
  endings are ignored

Risk Level: Low
Testing: Locally on Windows
Docs Changes: N/A
Release Notes: N/A
Fixes https://github.com/envoyproxy/envoy/issues/13191

### CODE REVIEW

**Comment by sunjayBhatia** [comment]
cc @envoyproxy/windows-dev

**Comment by sunjayBhatia** [comment]
Need to make binary mode open optional/refactor some tests

/wait

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc [hunk 1]
```diff
@@ -377,15 +377,15 @@ void testUtil(const TestUtilOptions& options) {
...
+            {{"\r", ""}, {"\n", "%0A"}, {" ", "%20"}, {"+", "%2B"}, {"/", "%2F"}, {"=", "%3D"}});
         // Assert twice to ensure a cached value is returned and still valid.
         EXPECT_EQ(urlencoded, server_connection->ssl()->urlEncodedPemEncodedPeerCertificate());
         EXPECT_EQ(urlencoded, server_connection->ssl()->urlEncodedPemEncodedPeerCertificate());
       }
       if (!options.expectedPeerCertChain().empty()) {
         std::string cert_chain = absl::StrReplaceAll(
             options.expectedPeerCertChain(),
-            {{"\n", "%0A"}, {" ", "%20"}, {"+", "%2B"}, {"/", "%2F"}, {"=", "%3D"}});
+            {{"\r", ""}, {"\n", "%0A"}, {" ", "%20"}, {"+", "%2B"}, {"/", "%2F"}, {"=", "%3D"}});
```
Comment (davinci26) [inline test/extensions/transport_sockets/tls/ssl_socket_test.cc:388]: nit: you can replace `Environment::NewLine` with `"%0A"`

**Comment by davinci26** [comment]
Looks good besides a small fix. Thanks a lot for getting this one to work!

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc (hunk 1 above)
Comment (sunjayBhatia) [reply test/extensions/transport_sockets/tls/ssl_socket_test.cc:388]: oh nice catch, thats a good call

**Comment by sunjayBhatia** [comment]
I believe the openssl fork errors are either caused by the Windows update service, or MSYS2 issues, hopefully addressed by one of: https://github.com/envoyproxy/envoy-build-tools/pull/101 or https://github.com/envoyproxy/envoy-build-tools/pull/100

**Comment by sunjayBhatia** [comment]
See https://github.com/envoyproxy/envoy/pull/13276

Looks like https://github.com/envoyproxy/envoy-build-tools/pull/100 is required to solve this issue in the build container/RBE

**Comment by wrowe** [comment]
/wait

on https://github.com/envoyproxy/envoy-build-tools/pull/100

**Comment by sunjayBhatia** [comment]
/wait

waiting on https://github.com/envoyproxy/envoy/pull/13367

**Comment by sunjayBhatia** [comment]
Failing due to same reason this test is being marked as flaky for the time being: https://github.com/envoyproxy/envoy/pull/13407

**Comment by sunjayBhatia** [comment]
/wait

still waiting on resolving the openssl msys2 fork issue, may continue with just checking in certs etc.

**Comment by sunjayBhatia** [comment]
Blocked on whichever https://github.com/envoyproxy/envoy/pull/13702 or https://github.com/envoyproxy/envoy-build-tools/pull/107 is chosen as the solution

**Comment by sunjayBhatia** [comment]
unblocked by https://github.com/envoyproxy/envoy/pull/13702

**Comment by sunjayBhatia** [comment]
@PiotrSikora this should be ready to go again, macos CI failure looks like one of the systematic ones we've been having generally

**Comment by sunjayBhatia** [comment]
merged master again to hopefully get the non-windows flakes out, @PiotrSikora have you had a chance to take a look at this yet?

**Review by PiotrSikora — APPROVED** [review]

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**